src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -650,6 +661,116 @@ vertexSet = Set.fromAscList . vertexList
+
+        runDfs :: Ord a => AM.AdjacencyMap a -> a -> PartMonad a (a, a, a)
+        runDfs g v = mod v <$> do m <- get
+                                  guard $ v `Map.notMember` m
+                                  dfs g LeftPart v
+
+        mod :: a -> (b, c) -> (a, b, c)
+        mod x (y, z) = (x, y, z)
+
+type PathMonad a = MaybeT (State (Set.Set a)) [a]
+
+findPath :: Ord a => a -> a -> AM.AdjacencyMap a -> Maybe [a]
+findPath u v g = evalState (runMaybeT $ dfs g v u) Set.empty
+    where
+        dfs :: Ord a => AM.AdjacencyMap a -> a -> a -> PathMonad a
+        dfs g t v | v == t = return [v]
+                  | otherwise = ((:) v) <$> do vis <- get
+                                               guard $ v `Set.notMember` vis
+                                               modify $ Set.insert v
+                                               asum $ map (dfs g t) $ neighbours v g
+
+-- | Test bipartiteness of given graph. In case of success, return an
+-- 'AdjacencyMap' with the same set of edges and each vertex marked with the
+-- part it belongs to. In case of failure, return any odd cycle in the graph.
+--
+-- The returned partition is lexicographicaly minimal. That is, consider the
+-- string of part identifiers for each vertex in ascending order. Then,
+-- considering that the identifier of the left part is less then the identifier
+-- of the right part, this string is lexicographically minimal of all such
+-- strings for all partitions.
+--
+-- /Note/: as 'AdjacencyMap' only represents __undirected__ bipartite graphs,
+-- all edges in the input graph are assumed to be bidirected and all edges in
+-- the output 'AdjacencyMap' are bidirected.
+--
+-- It is advised to use 'leftVertexList' and 'rightVertexList' to obtain the
+-- partition of the vertices and 'hasLeftVertex' and 'hasRightVertex' to check
+-- whether a vertex belongs to a part.
+--
+-- Complexity: /O((n + m) log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- detectParts 'Algebra.Graph.AdjacencyMap.empty'                                      == Right 'empty'
+-- detectParts ('Algebra.Graph.AdjacencyMap.vertex' x)                                 == Right ('leftVertex' x)
+-- detectParts (1 * (2 + 3))                              == Right ('edges' [(1, 2), (1, 3)])
+-- detectParts ((1 + 3) * (2 + 4) + 6 * 5)                == Right ('swap' (1 + 3) * (2 + 4) + 5 * 6)
+-- 'Data.Either.isRight' (detectParts ('Algebra.Graph.AdjacencyMap.edge' x y))                       == True
+-- 'Data.Either.isRight' (detectParts ('Algebra.Graph.AdjacencyMap.clique' x))                       == x < 3
+-- 'Data.Either.isRight' (detectParts ('Algebra.Graph.AdjacencyMap.star' x ys))                      == True
+-- 'Data.Either.isRight' (detectParts ('Algebra.Graph.AdjacencyMap.biclique' xs ys))                 == True
+-- 'Data.Either.isRight' (detectParts ('fromBipartite' ('toBipartite' am))) == True
+-- 'Data.Either.isRight' (detectParts (1 * 2 * 3))                      == False
```

Comment:
And, similarly, instead of `isRight ... == False` better give `== Left ...` examples instead.

Reply:
I don't see a formal property of the returned cycle we could guarantee to provide concrete examples in the docs. And even if the cycle is unique, we can't guarantee the order of vertices in it. For example, `detectParts (1 * 2 * 3)` might be equal to `Left [1, 2, 3]` or to `Left [2, 1, 3]` depending on the implementation. 
It is possible to provide some example like
```
 bimap sort id (detectParts (1 * 2 * 3)) == Left [1, 2, 3]
```
but this seems too complicated to me.

Reply:
Why not provide `detectParts (1 * 2 * 3) == Left [1, 2, 3]` if that's how the current implementation works? At least it will give the user a clue how the current implementation handles this particular case.

In general, it would of course be great if you could precisely describe the odd cycle that is obtained. Isn't it just "lexicographically least odd cycle"? If not, can you give a counterexample?

Reply:
For now, `detectParts (1 * 2 * 3) == Left [2, 3, 1]`.

There is a simple example where even the set of vertices in the returned odd cycle is not lexicographically minimal: 
```
detectParts (1 * (3 + 6) + 3 * 4 * 5 + 2 * 6 * 7) == Left [4, 5, 3]
```
But there is a cycle `[2, 6, 7]`. 

The problem is that the property "first odd cycle found by the depth-first search that goes through vertices in ascending order" doesn't seem to tell anything special at all. For instance, in the above example it goes `1 -> 3 -> 4 -> 5 -> 3` and doesn't even visit vertex 2.

Reply:
I added `detectParts (1 * 2 * 3) == Left [2, 3, 1]`.

Reply:
After discussion with @snowleopard we agreed on the following definition of optimality of the returned odd cycle:

> There exists a path that is either empty or ends in a vertex adjacent to the first vertex in the cycle, such that all vertices in (path ++ cycle) are distinct and (path ++ cycle) is lexicographically minimal among all such pairs.

For example, in `(1 * 3 * 4) + 2 * (1 + 2)` we will return the cycle `[2]` and the path is `[1]`. Note that we can't compare pairs (path, cycle), because the pair `([], [1, 3, 4])` is valid and lexicographically smaller than the pair `([1], [2])`.

So the example finally becomes `detectParts (1 * 2 * 3) == Left [1, 2, 3]`.